apps/b2c/pages/api/auth/[...nextauth].ts

```diff
@@ -0,0 +1,155 @@
+
+export default async function auth(req: NextApiRequest, res: NextApiResponse) {
+  // Do whatever you want here, before the request is passed down to `NextAuth`
+
+  return await NextAuth(req, res, {
+    adapter: HasuraAdapter(),
+    providers: [
+      EmailProvider({
+        server: process.env.EMAIL_SERVER!,
+        from: process.env.EMAIL_FROM!,
+      }),
+      GoogleProvider({
+        clientId: process.env.GOOGLE_CLIENT_ID!,
+        clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
+      }),
+      FacebookProvider({
+        clientId: process.env.FACEBOOK_CLIENT_ID!,
+        clientSecret: process.env.FACEBOOK_CLIENT_SECRET!,
+      }),
+      CredentialsProvider({
+        name: 'Credentials',
+        credentials: {
+          username: {
+            label: 'Username',
+            type: 'text',
+            placeholder: 'Username',
+          },
+          password: {
+            label: 'Password',
+            type: 'password',
+            placeholder: 'Password',
+          },
+        },
+        authorize: async (credentials) => {
+          const { username, password } = credentials || {}
+
+          const data = await asyncSSRFetch<{ authenticate: User }>({
+            query: GET_USER_WITH_CREDENTIALS,
+            variables: {
+              username,
+              password,
+            },
+          })
+
+          if (!data.authenticate.id) return null
+
+          const { id, name, role, email, image } = data.authenticate
+          return { id, name, role, email, image }
+        },
+      }),
+    ],
+    jwt: {
+      maxAge: MAX_AGE,
+      encode: async ({ token, secret, maxAge }) => {
+        token = {
+          ...token,
+          exp: getExp({ exp: token?.exp, maxAge }),
+        } as JWT
+
+        const encodedToken = jwt.sign(token, secret, { algorithm: 'HS256' })
+        return Promise.resolve(encodedToken)
+      },
+      decode: async ({ token, secret }) => {
+        if (!token) return null as Awaitable<null>
+
+        const decodedToken = jwt.verify(token, secret, {
+          algorithms: ['HS256'],
+        })
+        return decodedToken as Awaitable<JWT>
+      },
+    },
+    callbacks: {
+      jwt: async ({ token, user }) => {
+        if (user) {
+          const role = user.role ? user.role : 'user'
+
+          token = {
+            ...token,
+            'https://hasura.io/jwt/claims': {
+              'x-hasura-allowed-roles': [role],
+              'x-hasura-default-role': role,
+              'x-hasura-role': role,
+              ...(role === 'user'
+                ? { 'x-hasura-user-id': token.sub }
+                : { 'x-hasura-account-id': token.sub }),
+            },
+          }
+        }
+
+        return Promise.resolve(token)
+      },
+      session: async ({ session, token }) => {
+        const encodedToken = jwt.sign(token, process.env.NEXTAUTH_SECRET!, {
+          algorithm: 'HS256',
+        })
+
+        session.user = {
```

Comment:
can be more immutable ex:
```ts
const sessionResponse: User = {
  user: { //.. },
  token: { //.. }
}
```

not familiar with the api but it's nice to check if `callbacks.session` is already strictly typed and use those for inference rather that using `as User`

Reply:
updated